CH03-IOCoreSSL/CH03S01-Basic-NPN-and-ALPN.en.md

```diff
@@ -0,0 +1,199 @@
+# Basic Knowledge: NPN and APLN
+Before you read the implementation of SSL in ATS, Let's learn about NPN and APLN first. They are supposed to solve the same issue:
+  - Which protocol will be performed over plaintext after SSL connection is established?
+In the OSI model, SSL, widely used in many scenarios, is generally considered in the 6th layer. However,
+  - which protocol will be applied to the decrypted pipe, and how to apply it?
+  - It can only be decided by an agreement between the client and the server.
+  - For instance, on port 443, it's decided that we use SSL to encrypt traffic in HTTP/1.x.
+
+As SPDY is created, a requirement that using several possible application layer protocols on the same port has been put forward,
+
+  - We can't tell which protocol is applied unless data gets decrypted.
+  - We need to know which protocol before data gets decrypted.
+  - Therefore TLS-NPN extension is implemented.
+
+TLS-NPN extension implements that in SSL handshake:
+
+  - The server tells the client what protocols it supports.
+  - Then the client tells the server which protocol will be applied on this connection.
+
+TLS-NPN is short for Transport Layer Security - Next Protocol Negotiation (instead of NPN bipolar transistor, haha), it is implemented as an extension of TLS by Google to support SPDY as an application layer protocol.
+
+
+## TLS-NPN extension
+
+The detailed flow for a full SSL handshake with NPN is as follows. Please refer to [RFC 5246 Section 7.3](https://tools.ietf.org/html/rfc5246#section-7.3)
+
+```
+Client                                               Server
+
+ClientHello (NPN extension)   -------->
+                                               ServerHello
+                                                 (NPN extension &
+                                                  list of protocols)
+                                              Certificate*
+                                        ServerKeyExchange*
+                                       CertificateRequest*
+                            <--------      ServerHelloDone
+Certificate*
+ClientKeyExchange
+CertificateVerify*
+[ChangeCipherSpec]
+EncryptedExtensions
```

Comment:
Refer to https://tools.ietf.org/html/draft-agl-tls-nextprotoneg-04 [Page 03]
```
Therefore a full handshake with "EncryptedExtensions" has the
   following flow (contrast with section 7.3 of RFC 5246 [RFC5246])
```

```
The "EncryptedExtensions" message contains a series of "Extension"
   structures (see section 7.4.1.4 of RFC 5246 [RFC5246]

If the server included a "next_protocol_negotiation" extension in its
   "ServerHello" message, the client MUST include an "Extension" with
   "extension_type" equal to "next_protocol_negotiation(TBD)".  The
   "extension_data" of which has the following format:

   struct {
     opaque selected_protocol<0..255>;
     opaque padding<0..255>;
   } NextProtocolNegotiationEncryptedExtension;
```

Please check the chinese edition, I have a comment here.
```
EncryptedExtensions（包含NP信息）
EncryptedExtensions ( NP extension is included )
```